[TensorExpr] Add a tensor expressions tutorial.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45527 [TensorExpr] Add a tensor expressions tutorial.**



Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

Commit:
[WIP][TensorExpr] Add a tensor expressions tutorial.

[ghstack-poisoned]

Commit:
Update on "[WIP][TensorExpr] Add a tensor expressions tutorial."


Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

Commit:
Update on "[WIP][TensorExpr] Add a tensor expressions tutorial."




Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

Commit:
Update on "[WIP][TensorExpr] Add a tensor expressions tutorial."




Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a tensor expressions tutorial."




Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Add a tensor expressions tutorial."




Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a tensor expressions tutorial."




Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a tensor expressions tutorial."




Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

eellison commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,366 @@
+    For* j_tail;
+    loopnest.splitWithTail(
+        loops[1], // loops[0] is the outer loop, loops[1] is inner
+        9,
```
maybe give 9 a named variable

eellison commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,366 @@
+    // computation everything is ready.
+
+    // Let's now create some inputs and run our computation with them:
+    std::vector<int> data_A(64 * 32, 3); // This will be the input A
```
are strides / non-contiguity ever a thing?

ZolotukhinM replied:
We don't have strides/non-contiguity concepts at the moment, but @asuhan had some thoughts on whether we should have them.

asuhan replied:
@eellison We might need them, or not, depending on other decisions. Non-default strides can be used to lower `transpose` as a no-op in NNC. For example, for `x.transpose(0, 1)`, just permute the strides at positions 1 and 0. This will change address calculations for all subsequent operations on `x`, no data is moved. I'm not sure how well this will work in practice - for example, BERT calls `contiguous` right after `transpose`, so it'd defeat this trick. As of now we don't need much / anything from NNC to support this trick though, so it doesn't hurt to document it.

ngimel replied:
@mzolotukhin if you don't have non-contiguity concepts, does it mean that if someone passes discontiguous tensor as an input `.contiguous()` would need to be called on it first?

ZolotukhinM replied:
@ngimel, as of now, we will reject such value in the fuser pass since we're checking for complete static shapes there (and I believe/hope we check for contiguity). How we support that in a long term has not been decided yet, but if we don't want to represent strides, then - you are right - we will have to either require all inputs to be contiguous or call `.contiguous` on them in the fused kernel.

ngimel replied:
Do you guys have tests verifying that discontiguous inputs will be rejected?

ZolotukhinM replied:
I take back what I said about rejecting non-contiguous. We do support non-contiguous inputs, even though we don't have a concept representing strides in tensor expressions. The way it works is that we generate a valid index expression when we lower an access to a non-contiguous value:
https://github.com/pytorch/pytorch/blob/546aab66c1496845134538e78d6653fc4064ad63/torch/csrc/jit/tensorexpr/kernel.cpp#L1462-L1474

When you asked about the test I remembered that I wrote one exactly for that :)
https://github.com/pytorch/pytorch/blob/546aab66c1496845134538e78d6653fc4064ad63/test/cpp/tensorexpr/test_kernel.cpp#L102-L107

eellison commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,366 @@
+    //   Y = i + j
+    // }
+
+    // A Tensor refers to an individual computation defined by a Function. For
```
does a Function always return a new Tensor? i guess i'm a still a little confused abt the difference between a Function and a Block

ZolotukhinM replied:
> does a Function always return a new Tensor?

No, technically when we create a Function we don't create any Tensors. However, to create a Tensor one need to create a Function.

> i'm a still a little confused abt the difference between a Function and a Block
Function and Tensor are very close concepts, but Function and Block are very different. Function here is not a function in programming languages sense (which can be also called method/procedure in some context). Here Function is a function in a mathematical sense, i.e. it's a formula, something like f(x) = x^2.

mruberry replied:
@ngimel and I also find this comment to be unclear.

eellison reviewed (approved):
Awesome, thanks for doing this! This is really clear & thorough, and great for ramping up on the code base! We might consider lifting some of these comments onto the more sparsely-commented code base.

Commit:
Update on "[TensorExpr] Add a tensor expressions tutorial."




Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a tensor expressions tutorial."




Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a tensor expressions tutorial."




Differential Revision: [D23998787](https://our.internmc.facebook.com/intern/diff/D23998787)

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,366 @@
+    // computations, and in order to represent accesses to tensors data, there
+    // is a special kind of expression - a load.
+    // To construct a load we need two pieces: the base and the indices. The
+    // base of a load is a Buf expression, which could be thought of as a
```
`Buf` are symbols representing memory domains, they are not real buffers (the name `Buf` might be poor btw and we're considering changing it). `Buf` is similar to `Var`, but it also specifies dimensions (`Var`s represent scalar variables that don't have dimensions).

ZolotukhinM commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,366 @@
+        new Function({"X", "Y"}, dims, args, {func_body1, func_body2});
+    std::cout << "Tensor function: " << *func << std::endl;
+    // Prints:
+    // Tensor function: Function F(i[64], j[32]) {
```
Tried to expand on it a little. The purpose of the Function is to "bundle" several tensor computations together. An example of it can be something along the lines of finding maximum element of an array and its index at the same time - these would be two different computations performed over the same memory domain that we might want to bundle together in a single Function. It is a bit hard to be super-specific with the example though because currently our code for supporting multi-output functions is not done and we can't really have multi-output functions. One-output function is essentially the same as Tensor.

ZolotukhinM commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,366 @@
+        {{64, "i"}, {32, "j"}},
+        [](const VarHandle& i, const VarHandle& j) { return i / j; });
+    std::cout << "Tensor computation: " << *Z << std::endl;
+    // Prints: Tensor computation: Tensor Z(i[64], j[32]) = i / j
```
Our story with Functions is not complete. Currently we mostly expect users (~whoever uses `Compute` API) to operate on Tensors and Placeholders. So far there has been no need in exposing Functions there - but maybe it'll change in future (e.g. if we properly support multi-output functions).

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    // operations and allow us to express the same semantics in a more natural
+    // way:
+    ExprHandle l = 1;
+    ExprHandle r = Var::make("x", kInt);
```
Does the concept of a "Var" need to be introduced previously? Also, it may be confusing for readers to see an expression with a variable not assigned a value at this point.

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    ExprHandle b = Var::make("b", kFloat);
+    ExprHandle c = Var::make("c", kFloat);
+    ExprHandle x = ExprHandle(5) * a + b / (sigmoid(c) - 3.0f);
+    std::cout << "Tensor expression: " << *x.node() << std::endl;
```
How does type promotion work?

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    // computations, and in order to represent accesses to tensors data, there
+    // is a special kind of expression - a load.
+    // To construct a load we need two pieces: the base and the indices. The
+    // base of a load is a Buf expression, which could be thought of as a
```
This introduces a lot of new concepts at once: "Buf"s, bases, indices, loads, dimensions.

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    //
+    // Let's construct a simple load:
+    BufHandle A("A", {ExprHandle(64), ExprHandle(32)}, kInt);
+    ExprHandle i = Var::make("i", kInt), j = Var::make("j", kInt);
```
Suggest one definition (or declaration) per line.

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    // Let's construct a simple load:
+    BufHandle A("A", {ExprHandle(64), ExprHandle(32)}, kInt);
+    ExprHandle i = Var::make("i", kInt), j = Var::make("j", kInt);
+    ExprHandle load = Load::make(A.dtype(), A, {i, j}, /* mask= */ 1);
```
Mask value may be confusing here since it wasn't previously mentioned.

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    // base of a load is a Buf expression, which could be thought of as a
+    // placeholder similar to Var, but with dimensions info.
+    //
+    // Let's construct a simple load:
```
It may be helpful to describe what the snippet will do upfront so it's easier for a reader to follow along.

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    ExprHandle i = Var::make("i", kInt), j = Var::make("j", kInt);
+    ExprHandle load = Load::make(A.dtype(), A, {i, j}, /* mask= */ 1);
+    std::cout << "Tensor expression: " << *load.node() << std::endl;
+    // Prints: Tensor expression: A[i, j]
```
The result here doesn't really connect with the motivation for this snippet, which was to learn about loading data. The snippet only demonstrates the syntax to construct this printed syntax.

Does this actually show how to do a load? Or is this just syntax to logically reference the elements in the buffer?

Follow-up question: is there not sugar to do A[i, j] directly?

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+
+  std::cout << "*** Tensors, Functions, and Placeholders ***" << std::endl;
+  {
+    // A tensor computation is represented by objects of Tensor class and
```
Is a tensor computation != a tensor expression?

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    // First, we need to specify the domain, or dimensions in which the
+    // computation would be performed. Let's create a 64x32 domain:
+    std::vector<const Expr*> dims = {
+        new IntImm(64), new IntImm(32)}; // IntImm stands for Integer Immediate
```
This comment seems like it should be on the first instance of IntImm.

gchanan replied:
on the first example I guessed it meant "Integer Immutable."

ZolotukhinM replied:
Good catch, thanks! @gchanan, you were not alone with such a guess :)

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    // expressions for each computation (which can be accessed via
+    // `func->func_var(idx)`) with the names specified by the first parameter of
+    // the constructor call. In our example two `Buf` variables will be created
+    // with names 'X' and 'Y', each of them would signify a domain of 64x32.
```
So dims describes the dimensionality of ALL the variables in this case?

mruberry replied:
Or, following the later part of this, is it that each elements of dims is assigned to the corresponding element of the args array?

mruberry replied:
The phrase "would signify a domain of 64x32" does not seem to match what's printed below "i[64], j[32]"?

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    // We can now print out our function:
+    std::cout << "Tensor function: " << *func << std::endl;
+    // Prints:
+    // Tensor function: Function F(i[64], j[32]) {
```
The use of "i" and "j" here is confusing. Is this saying that i is a 1D array of 64 elements and j is a 1D array of 32 elements? What does it mean to multiply (or add) two such arrays?

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+
+    // A Tensor refers to an individual computation defined by a Function. For
+    // instance, we could create a following tensor given the function above:
+    int output_idx = 0; // Used to index the computation
```
Can you elaborate on what this "indexing" does?

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    Tensor* Z = Compute(
+        "Z",
+        {{64, "i"}, {32, "j"}},
+        [](const VarHandle& i, const VarHandle& j) { return i / j; });
```
This seems like an "i" and "j" name collision?

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    std::cout << "Tensor computation: " << *Z << std::endl;
+    // Prints: Tensor computation: Tensor Z(i[64], j[32]) = i / j
+
+    // Tensors might access other tensors and external placeholders in their
```
While the term "placeholder" has been mentioned informally above, it hasn't been formally defined yet

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+        "R",
+        {{64, "i"}, {32, "j"}},
+        [&](const VarHandle& i, const VarHandle& j) {
+          return Z->call(i, j) * P.load(i, j);
```
call() has not yet been introduced

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    //
+    // Why do we have Functions and Tensors and what is the relationship between
+    // them? Functions are used to represent several computations performed over
+    // the same domain. Tensors refer to individual computations of a Function.
```
An example illustrating this would be helpful, since that doesn't seem to be what's shown above.

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+
+    // Creating a loop nest is as quite simple, we just need to specify what are
+    // the output tensors in our computation and LoopNest object will
+    // automatically pull all tensor dependencies:
```
Would the reader understand what "pull[ing] all tensor dependencies" means or why it's important?

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    // element, and the value to be stored, which is an arbitrary expression.
+
+    // Once we've constructed the loop nest, we can apply various tranformations
+    // to it. To begin with, let's inline computation of X into computation of Y
```
Maybe describe what inlining is first?

mruberry commented on test/cpp/tensorexpr/tutorial.cpp:
```diff
@@ -0,0 +1,426 @@
+    // Let's now create some inputs and run our computation with them:
+    std::vector<int> data_A(64 * 32, 3); // This will be the input A
+    std::vector<int> data_B(64 * 32, 5); // This will be the input B
+    std::vector<int> data_X(64 * 32, 0); // This will be used for the result
```
Maybe don't initialize data_X to any value since it will be written to?

ZolotukhinM commented:
Thanks for the comments, I'll try to address them in a follow up!
